src/rbindlist.c

```diff
@@ -514,7 +514,16 @@ SEXP rbindlist(SEXP l, SEXP usenamesArg, SEXP fillArg, SEXP idcolArg)
         if (w==-1 || !length(thisCol=VECTOR_ELT(li, w))) {  // !length for zeroCol warning above; #1871
           writeNA(target, ansloc, thisnrow);  // writeNA is integer64 aware and writes INT64_MIN
         } else {
-          if ((TYPEOF(target)==VECSXP || TYPEOF(target)==EXPRSXP) && TYPEOF(thisCol)!=TYPEOF(target)) {
+          if ((TYPEOF(target)==VECSXP) && TYPEOF(thisCol)>TYPEOF(target)) {
+            // Exotic non-atomic types need each element to be wrapped in a list, e.g. expression vectors #546
+            SEXP thisColList = PROTECT(allocVector(VECSXP, length(thisCol))); nprotect++;
+            for(int r=0; r<length(thisCol); ++r) {
+              SEXP thisElement = VECTOR_ELT(thisCol, r);
+              if (TYPEOF(thisCol) == EXPRSXP) thisElement = coerceVector(thisElement, EXPRSXP); // otherwise LANGSXP
```

Comment:
Do we need another PROTECT call here or on the line before?

Reply:
We need. Coerced object is new memory location not referenced anywhere this can be gc.